```
Relibank Test Suite — sandbox (branch: ftr/scenario-ui-flag)
GitHub Actions run: https://github.com/newrelic/relibank/actions/runs/28979937323
target_environment=sandbox   test_suite=all   effective color=blue (auto-resolved)
================================================================================

SUMMARY
  Python modules: 15 passed, 2 failed (out-of-scope New Relic latency/ingestion bucket)
  Frontend: 37 passed (5 files)

PYTHON TESTS (per module)
--------------------------------------------------------------------------------
Running test_ab_testing_scenarios.py...
============================= 10 passed in 19.77s ==============================
✓ test_ab_testing_scenarios.py passed
Running test_apm_user_tracking.py...
============================= 15 passed in 22.89s ==============================
✓ test_apm_user_tracking.py passed
Running test_browser_user_tracking.py...
============================== 7 passed in 4.17s ===============================
✓ test_browser_user_tracking.py passed
Running test_card_payment_transactions.py...
========================= 4 passed in 77.13s (0:01:17) =========================
✓ test_card_payment_transactions.py passed
Running test_db_pool_e2e.py...
======================== 1 passed in 144.92s (0:02:24) =========================
✓ test_db_pool_e2e.py passed
Running test_db_pool_scenarios.py...
============================== 7 passed in 8.22s ===============================
✓ test_db_pool_scenarios.py passed
Running test_end_to_end.py...
============================= 23 passed in 39.50s ==============================
✓ test_end_to_end.py passed
Running test_mfe_single.py...
======================== 1 passed in 225.58s (0:03:45) =========================
✓ test_mfe_single.py passed
Running test_microfrontend_telemetry.py...
======================== 4 passed in 504.45s (0:08:24) =========================
✓ test_microfrontend_telemetry.py passed
Running test_newrelic_instrumentation.py...
=================== 5 passed, 1 skipped in 184.88s (0:03:04) ===================
✓ test_newrelic_instrumentation.py passed
Running test_newrelic_risk_assessment.py...
======================== 4 failed in 193.77s (0:03:13) =========================
✗ test_newrelic_risk_assessment.py failed
Running test_nrdot_mssql_collector.py...
======================== 5 errors in 300.10s (0:05:00) =========================
✗ test_nrdot_mssql_collector.py failed
Running test_payment_scenarios.py...
======================== 5 passed in 129.26s (0:02:09) =========================
✓ test_payment_scenarios.py passed
Running test_rogue_deployment_scenarios.py...
============================== 7 passed in 49.86s ==============================
✓ test_rogue_deployment_scenarios.py passed
Running test_scenario_service.py...
============================== 16 passed in 5.73s ==============================
✓ test_scenario_service.py passed
Running test_stress_scenarios.py...
============================== 1 passed in 4.56s ===============================
✓ test_stress_scenarios.py passed
Running test_with_interactions.py...
======================== 2 passed in 252.07s (0:04:12) =========================
✓ test_with_interactions.py passed

FAILURE DETAIL — the 2 failing modules are the known, out-of-scope bucket
--------------------------------------------------------------------------------
FAILED test_newrelic_risk_assessment.py::test_risk_assessment_logs_in_newrelic - TypeError: 'NoneType' object is not subscriptable
FAILED test_newrelic_risk_assessment.py::test_declined_payment_shows_in_newrelic - AssertionError: Expected declined payment logs, found 0
FAILED test_newrelic_risk_assessment.py::test_agent_model_tracking_in_newrelic - AssertionError: Expected agent model tracking in logs, found 0
FAILED test_newrelic_risk_assessment.py::test_risk_assessment_ebpf_traces - TypeError: unsupported format string passed to NoneType.__format__
ERROR test_nrdot_mssql_collector.py::test_lock_contention_metrics - Failed: Timeout (>300.0s) from pytest-timeout.
ERROR test_nrdot_mssql_collector.py::test_connection_and_throughput_metrics - Failed: Timeout (>300.0s) from pytest-timeout.
ERROR test_nrdot_mssql_collector.py::test_buffer_pool_metrics - Failed: Timeout (>300.0s) from pytest-timeout.
ERROR test_nrdot_mssql_collector.py::test_sql_compilation_metrics - Failed: Timeout (>300.0s) from pytest-timeout.
ERROR test_nrdot_mssql_collector.py::test_wait_stats_metrics - Failed: Timeout (>300.0s) from pytest-timeout.

FRONTEND TESTS
--------------------------------------------------------------------------------
Test Files  5 passed (5)
      Tests  37 passed (37)
```

Definitive confirmation — these failures are pre-existing on main, not introduced by this branch. Both modules fail identically on:

main @ 2026-07-02 (run 28594218643) — before this branch existed (branch's first commit was ~July 6)
main @ 2026-07-09 (run 29023337246) — current main

--- a/.github/workflows/deploy-relibank.yml
+++ b/.github/workflows/deploy-relibank.yml
@@ -275,3 +275,6 @@ jobs:
     with:
       target_environment: ${{ github.event.inputs.environment }}
       test_suite: all
+      # Validate the just-deployed color before cutover (via X-Test-Env). The scenario-UI
+      # expected state is derived from the environment inside test-suite.yml (prod -> off).
+      target_color: ${{ github.event.inputs.target_color }}

--- a/.github/workflows/test-suite.yml
+++ b/.github/workflows/test-suite.yml
@@ -29,6 +29,15 @@ on:
           - payment
           - frontend
           - smoke
+      target_color:
+        description: 'Color to direct scenario-UI checks at (sends X-Test-Env). Empty = active color.'
+        required: false
+        default: ''
+        type: choice
+        options:
+          - ''
+          - blue
+          - green
   workflow_call:
     inputs:
       target_environment:
@@ -38,6 +47,10 @@ on:
         type: string
         required: false
         default: 'all'
+      target_color:
+        type: string
+        required: false
+        default: ''
 
 jobs:
   python-tests:
@@ -95,6 +108,47 @@ jobs:
           cd tests
           pip install -r requirements.txt
 
+      # Terraform-deployed envs deploy blue/green and keep MSSQL as a headless ClusterIP (no
+      # public endpoint). Authenticate to AKS, resolve the color under test, and port-forward
+      # THAT color's mssql to 127.0.0.1:1433 for the DB-direct tests. The resolved color is
+      # exported as EFFECTIVE_COLOR so the HTTP tests route to it (X-Test-Env, via conftest) and
+      # NR queries can filter by it. Skipped for legacy envs (events) with no AKS vars — those
+      # are single-namespace (no color) and use their own DB_SERVER secret.
+      - name: Azure Login (color-aware DB tunnel)
+        if: ${{ vars.AKS_CLUSTER_NAME != '' }}
+        uses: azure/login@v2
+        with:
+          creds: ${{ secrets.AZURE_CREDENTIALS }}
+
+      - name: Resolve effective color
+        if: ${{ vars.AKS_CLUSTER_NAME != '' }}
+        run: |
+          az aks get-credentials \
+            --resource-group "${{ vars.AKS_RESOURCE_GROUP }}" \
+            --name "${{ vars.AKS_CLUSTER_NAME }}" \
+            --overwrite-existing
+          # Directed color when provided (deploy/manual); else the ACTIVE color from
+          # main-ingress (cron/no-input). blue-service|green-service -> blue|green.
+          color="${{ inputs.target_color }}"
+          if [ -z "$color" ]; then
+            backend=$(kubectl get ingress main-ingress -n default \
+              -o jsonpath='{.spec.rules[0].http.paths[0].backend.service.name}')
+            color="${backend%-service}"
+          fi
+          echo "Effective color: ${color:-<none>}"
+          echo "EFFECTIVE_COLOR=${color}" >> "$GITHUB_ENV"
+
+      - name: Port-forward MSSQL (effective color)
+        if: ${{ vars.AKS_CLUSTER_NAME != '' }}
+        run: |
+          kubectl port-forward -n "relibank-${EFFECTIVE_COLOR}" svc/mssql 1433:1433 >/tmp/mssql-pf.log 2>&1 &
+          # Wait until 127.0.0.1:1433 accepts connections (~30s max)
+          for i in $(seq 1 30); do
+            (exec 3<>/dev/tcp/127.0.0.1/1433) 2>/dev/null && { echo "MSSQL tunnel up (relibank-${EFFECTIVE_COLOR})"; exec 3>&- ; break; }
+            sleep 1
+          done
+          (exec 3<>/dev/tcp/127.0.0.1/1433) 2>/dev/null || { echo "port-forward failed:"; cat /tmp/mssql-pf.log; exit 1; }
+
       - name: Run all Python tests individually
         if: ${{ inputs.test_suite == 'all' || inputs.test_suite == '' }}
         id: individual-tests
@@ -110,8 +164,18 @@ jobs:
           SCENARIO_SERVICE_URL: ${{ vars.BASE_URL }}
           NEW_RELIC_USER_API_KEY: ${{ secrets.NR_USER_API_KEY }}
           NEW_RELIC_ACCOUNT_ID: ${{ vars.NR_ACCOUNT_ID }}
-          DB_SERVER: ${{ secrets.DB_SERVER }}
-          DB_PASSWORD: ${{ secrets.DB_PASSWORD }}
+          # Scenario-UI webpage test: direct at a color (X-Test-Env) + expected flag state,
+          # threaded from the deploy so no manual sync. Empty color = active color.
+          TARGET_COLOR: ${{ env.EFFECTIVE_COLOR || inputs.target_color }}
+          SCENARIO_UI_ENABLED: ${{ inputs.target_environment == 'prod' && 'false' || 'true' }}
+          # DB creds are convention-agnostic and mostly zero-config:
+          #  - host: legacy events has a DB_SERVER secret (public LB) and it wins; Terraform envs
+          #    have none, so tests hit 127.0.0.1 — the "Port-forward MSSQL" step below tunnels the
+          #    active color's in-cluster (ClusterIP) DB to localhost:1433. No public exposure.
+          #  - auth: reuse the MSSQL_SA_* secrets Terraform envs already have (events' DB_PASSWORD wins).
+          DB_SERVER: ${{ secrets.DB_SERVER || '127.0.0.1' }}
+          DB_USERNAME: ${{ secrets.DB_USERNAME || secrets.MSSQL_SA_USER }}
+          DB_PASSWORD: ${{ secrets.DB_PASSWORD || secrets.MSSQL_SA_PASSWORD }}
         run: |
           cd tests
 
@@ -163,6 +227,9 @@ jobs:
           SCENARIO_SERVICE_URL: ${{ vars.BASE_URL }}
           NEW_RELIC_USER_API_KEY: ${{ secrets.NR_USER_API_KEY }}
           NEW_RELIC_ACCOUNT_ID: ${{ vars.NR_ACCOUNT_ID }}
+          # Scenario-UI webpage test: X-Test-Env color + expected flag (empty color = active)
+          TARGET_COLOR: ${{ env.EFFECTIVE_COLOR || inputs.target_color }}
+          SCENARIO_UI_ENABLED: ${{ inputs.target_environment == 'prod' && 'false' || 'true' }}
         run: |
           cd tests
           pytest test_scenario_service.py --tb=line

--- a/README.md
+++ b/README.md
@@ -45,6 +45,7 @@ Relibank simulates a banking system with separate services for accounts, transac
     - AI agent configuration for risk assessment (normal vs rogue agent)
     - Chaos Mesh experiment triggers
     - Locust load testing integration
+    - Web UI visibility derived from the environment (hidden in `prod`, shown elsewhere); the API stays reachable regardless
 
 ## Getting Started
 

--- a/docs/deployer/runbook.md
+++ b/docs/deployer/runbook.md
@@ -154,6 +154,14 @@ Recommended: open a small patch PR off `main` that adds the new value to every l
 4. Re-run `Test Suite` workflow to verify post-flip.
 5. **`Deploy ReliBank`** — `action_type=destroy`, `target_color=blue` (cleanup; not strictly required, but reclaims node pool capacity).
 
+### Scenario Runner UI visibility
+
+Whether the browser page at `/scenario-runner/home` is served (200) or hidden (404) is **derived from the environment — not a workflow input**: **`prod` → hidden, every other environment → shown**. The scenario API (`/scenario-runner/api/*`) is reachable either way, so backend consumers and automation keep working regardless.
+
+The rule lives in `terraform/aks/app_module/main.tf` — the `infrastructure-config` ConfigMap sets `SCENARIO_UI_ENABLED = tostring(var.demo_environment != "prod")`, baked into the scenario-runner pod at startup. So it takes effect on a normal `Deploy ReliBank` of the target environment/color; there's nothing to toggle (and no per-run flag to remember).
+
+**Automated verification.** `tests/test_scenario_service.py::test_scenario_ui_flag_controls_webpage` asserts the page matches the expected state: `/scenario-runner/home` returns 200 (with the page marker) when enabled, else 404, and `/scenario-runner/api/scenarios` stays 200 either way. It's **color-directed** — set `TARGET_COLOR` to send `X-Test-Env: <color>` and validate a specific color (empty = active). The expected state is derived the same way as the deploy: `test-suite.yml` computes `SCENARIO_UI_ENABLED` from `target_environment` (`prod → false`, else `true`), so post-deploy runs validate the just-deployed color **before** cutover with no flag to pass.
+
 ### Infra-only refresh
 
 When AOAI models change or the Function App code/config needs an update — no need to touch the app tier.
@@ -266,6 +274,78 @@ kubectl rollout restart deployment -n relibank-{color}
 
 Fix: bump `MSSQL_MEMORY_LIMIT_MB` from the prod default of 1024 MB to 6144 MB. Also bump k8s resources (`requests: 4Gi/1cpu`, `limits: 8Gi/2cpu`). Demo loadgen for QPM panels: `utils/scripts/mssql/loadgen/db-direct/run-banking-load.sh` (~25s spending-velocity runs against 2M `BankTransactions`).
 
+### MSSQL out of disk: RelibankDB log filled the data volume
+
+**Symptoms:** The `mssql-0` pod is `Running` but its **log** shows `Error: 1105`/`Error: 1101` (can't allocate space) and sometimes `Operating system error 31` on a tempdb file; `df -h /var/opt/mssql` in the pod is ~99% full. The **in-cluster app** starts failing DB writes (payment/ledger inserts hang or error). Note: this is distinct from the CI `pyodbc … Login timeout expired` symptom, which is a *reachability* problem (see [DB-direct tests can't reach the database from CI](#db-direct-tests-cant-reach-the-database-from-ci)), not a disk problem — a full disk shows up as 1105/1101 in the pod log, not as a connection timeout on the runner.
+
+**Root cause:** `RelibankDB` shipped in **FULL** recovery model with no log-backup job, so the transaction log (`RelibankDB_log.ldf`) grows without bound and fills the (historically 2Gi) PVC. Once the volume is full, SQL Server can't allocate pages and every connection times out.
+
+**Diagnose** (per color — repeat for the active one, and green if you may cut back to it):
+
+```bash
+ns=relibank-blue            # or relibank-green
+SA=$(kubectl get secret database-credentials -n $ns -o jsonpath='{.data.MSSQL_SA_PASSWORD}' | base64 -d)
+SQLCMD=/opt/mssql-tools18/bin/sqlcmd
+
+kubectl exec -n $ns mssql-0 -- df -h /var/opt/mssql          # is it ~99% full?
+kubectl exec -n $ns mssql-0 -- sh -c 'du -sh /var/opt/mssql/data/*.ldf'   # log file size
+kubectl exec -n $ns mssql-0 -- "$SQLCMD" -S localhost -U sa -P "$SA" -C \
+  -Q "SELECT recovery_model_desc, log_reuse_wait_desc FROM sys.databases WHERE name='RelibankDB';"
+```
+
+**Fix — step by step.** Shrink first (frees space with no restart), then expand the volume:
+
+1. Switch to SIMPLE recovery and shrink the log (relieves the disk immediately):
+
+   ```bash
+   kubectl exec -n $ns mssql-0 -- "$SQLCMD" -S localhost -U sa -P "$SA" -C -Q "
+     ALTER DATABASE RelibankDB SET RECOVERY SIMPLE;
+     USE RelibankDB; CHECKPOINT;
+     DBCC SHRINKFILE (RelibankDB_log, 256);"
+   kubectl exec -n $ns mssql-0 -- df -h /var/opt/mssql        # should drop well below 100%
+   ```
+
+   > If `log_reuse_wait_desc` is `OLDEST_PAGE` the log may refuse to shrink until the pod is restarted (step 3). Re-run this `CHECKPOINT; DBCC SHRINKFILE` after the restart and it will drop to ~256 MB.
+
+2. Expand the PVC to 8Gi (storageclass `disk.csi.azure.com` has `allowVolumeExpansion=true`):
+
+   ```bash
+   kubectl patch pvc mssql-mssql-0 -n $ns -p '{"spec":{"resources":{"requests":{"storage":"8Gi"}}}}'
+   ```
+
+3. Finalize the resize — Azure disk grows the filesystem only on remount, so restart the pod once. **This is a brief DB outage; on the active color expect ~1 min where the app can't reach the DB.**
+
+   ```bash
+   kubectl delete pod mssql-0 -n $ns
+   kubectl rollout status statefulset/mssql -n $ns --timeout=180s
+   kubectl exec -n $ns mssql-0 -- df -h /var/opt/mssql        # now ~8Gi
+   ```
+
+4. Verify:
+
+   ```bash
+   kubectl exec -n $ns mssql-0 -- "$SQLCMD" -S localhost -U sa -P "$SA" -C \
+     -Q "SELECT recovery_model_desc, state_desc FROM sys.databases WHERE name='RelibankDB';"   # SIMPLE / ONLINE
+   ```
+
+**Prevention (already in IaC):** the PVC template is now `8Gi` (`terraform/aks/app_module/main.tf` `volume_claim_template`, mirrored in `k8s/base/databases/mssql-deployment.yaml`) and `init.sql` sets `RECOVERY SIMPLE` right after `CREATE DATABASE`, so new environments get both automatically.
+
+> ⚠️ **Applying the 8Gi template to an *existing* env replaces the StatefulSet.** A `volume_claim_template` change is immutable, so `terraform apply` will destroy/recreate the `mssql` StatefulSet. StatefulSet PVCs are retained on delete (not garbage-collected), so **data survives** and the new STS re-adopts the existing `mssql-mssql-0` PVC — but the pod restarts. For an already-running env you've patched by hand (steps above), the live PVC is already 8Gi, so this is a no-data-loss pod restart; schedule it, or `terraform state rm kubernetes_stateful_set_v1.mssql` and re-import to avoid the churn.
+
+### DB-direct tests can't reach the database from CI
+
+**Symptoms:** `test_card_payment_transactions` and `test_nrdot_mssql_collector` fail on *every* run (sandbox, staging, analysts) with `pyodbc.OperationalError ('HYT00', ... Login timeout expired)`. HTTP-only tests pass. This is **not** DB health or warm-up — the DB is simply unreachable from the runner.
+
+**Root cause:** those tests open a **direct** SQL connection to `secrets.DB_SERVER:1433`. That only works if MSSQL has a public endpoint. Old prod/`events` was deployed from `k8s/base`, where the mssql Service is `type: LoadBalancer` (`mssql-0`, public IP `relibank-mssql.westus2.cloudapp.azure.com`). The newer **Terraform blue/green deployer** (`app_module`) defines mssql as a **headless ClusterIP** (`cluster_ip = "None"`) with no external IP, so a GitHub-hosted runner has no route to it. HTTP tests still pass because they go through the public ingress LB (`{env}.{dns_zone}`), not the DB.
+
+**Fix — the test suite tunnels to the DB privately; MSSQL stays ClusterIP (matches demogorgon; no public exposure).** `test-suite.yml`'s `python-tests` job, when the target env has AKS vars (`vars.AKS_CLUSTER_NAME != ''`), logs into Azure (`azure/login@v2` + `secrets.AZURE_CREDENTIALS`), runs `az aks get-credentials`, detects the **active color** from `main-ingress`, and `kubectl port-forward`s that color's `svc/mssql` to `127.0.0.1:1433`. Tests then connect to `127.0.0.1` (`DB_SERVER: ${{ secrets.DB_SERVER || '127.0.0.1' }}`), authenticating with the `MSSQL_SA_*` secrets each Terraform env already has (`DB_PASSWORD: ${{ secrets.DB_PASSWORD || secrets.MSSQL_SA_PASSWORD }}`, `DB_USERNAME: ${{ secrets.DB_USERNAME || secrets.MSSQL_SA_USER }}`).
+
+**Zero manual setup for Terraform envs** — no `DB_SERVER`/`DB_PASSWORD` variables or secrets to create. It only requires what a deployable env already has: `AKS_CLUSTER_NAME` / `AKS_RESOURCE_GROUP` **variables** and the `AZURE_CREDENTIALS` **secret**. Just re-run **Relibank Test Suite** (`target_environment={env}`) and `test_card_payment_transactions` / `test_nrdot_mssql_collector` connect over the tunnel.
+
+- **Color is auto-detected**, so it never appears in a hostname or variable — the tunnel just targets the active color's namespace. (Assumes tests run against the live color via `BASE_URL`; add a color override input if canary DB testing of the inactive color is ever needed.)
+- **Legacy `events`** has no AKS vars, so the tunnel steps skip and it keeps using its own `DB_SERVER`/`DB_PASSWORD` secrets (which win the `||`) against its public `k8s/base` LoadBalancer.
+- The pyodbc connection string already sets `TrustServerCertificate=yes`, so connecting to `127.0.0.1` (cert CN won't match) is fine.
+
 ### Infra destroy times out with `context deadline exceeded` on `relibank` namespace
 
 Symptom: Stage 2 (`Terraform Destroy (Stage 2 infra)`) hangs for ~5min with `kubernetes_namespace_v1.relibank: Still destroying...`, then fails with `Error: context deadline exceeded`. `kubectl get ns relibank` shows status `Terminating` and `kubectl get stresschaos -n relibank -o jsonpath='{.items[*].metadata.finalizers}'` returns `["chaos-mesh/records"]`.
@@ -331,7 +411,7 @@ The Function App is **env-scoped** — both colors share one URL. After `stage=n
 
 ### Card payment / payment scenario tests intermittently fail on sandbox
 
-Likely Stripe rate-limit or green-color warm-up window (cold caches, JIT-compiled paths still warming). Re-run after a few minutes. Treat as a real failure only if persistent across multiple runs.
+Likely Stripe rate-limit or green-color warm-up window (cold caches, JIT-compiled paths still warming). Re-run after a few minutes. **If `test_card_payment_transactions` / `test_nrdot_mssql_collector` fail *every* run with `pyodbc … Login timeout expired`, that is not warm-up — those tests open a direct DB connection the CI runner can't reach on Terraform-deployed (blue/green) environments. See [DB-direct tests can't reach the database from CI](#db-direct-tests-cant-reach-the-database-from-ci).** (A separate failure mode — a genuinely full DB volume — shows up as `1105`/`1101` in the `mssql-0` pod log; see [MSSQL out of disk](#mssql-out-of-disk-relibankdb-log-filled-the-data-volume).)
 
 ### `db_pool_e2e` passes sandbox, fails prod
 

--- a/docs/deployer/testing-runbook.md
+++ b/docs/deployer/testing-runbook.md
@@ -1,0 +1,107 @@
+# ReliBank Testing Runbook
+
+Operational guide for running the **Relibank Test Suite** against a deployed environment — how to
+target a color, what the suite needs, and how to read the results. For the test catalog and local
+setup, see [`tests/README.md`](../../tests/README.md); for deploy/cutover mechanics, see
+[runbook.md](runbook.md).
+
+---
+
+## The workflow
+
+`Relibank Test Suite` ([`.github/workflows/test-suite.yml`](../../.github/workflows/test-suite.yml))
+runs the Python suite (per-module) plus the frontend (Vitest) suite. It fires three ways:
+
+| Trigger | Color / env | Notes |
+|---|---|---|
+| **Scheduled cron** (daily) | `events` env, **active** color | No inputs; validates whatever's live. |
+| **Manual** (`workflow_dispatch`) | you pick `target_environment` + optional `target_color` + `test_suite` | Use to test a specific env/color on demand. |
+| **Post-deploy** (`workflow_call` from `Deploy ReliBank`) | the **just-deployed** color | Runs automatically after a successful `action_type=deploy`, before cutover. |
+
+Inputs: `target_environment` (events/sandbox/staging/prod/analysts), `test_suite`
+(all/e2e/scenario/payment/frontend/smoke), `target_color` (blue/green/empty).
+
+---
+
+## Color-aware behavior
+
+Deployed environments are blue/green; the suite validates a **specific color** end to end. The
+*effective color* is resolved once in the `python-tests` job:
+
+1. `target_color` input if provided (post-deploy passes the just-deployed color), **else**
+2. the **active** color read from `main-ingress` (`kubectl get ingress main-ingress -n default`).
+
+That color drives all three layers:
+
+- **HTTP** — [`tests/conftest.py`](../../tests/conftest.py) sets `X-Test-Env: <color>` on every request
+  (canary ingress routes it to that color). Empty color → active.
+- **DB** — the job authenticates to AKS and `kubectl port-forward`s that color's `mssql` to
+  `127.0.0.1:1433`; DB-direct tests use `DB_SERVER=127.0.0.1`. MSSQL is never publicly exposed.
+- **New Relic** — [`tests/nrql_color.py`](../../tests/nrql_color.py) appends a color filter to NRQL:
+  `Transaction`→`deploy.color`, `Span`→`k8s.namespace.name`, `Log`→`namespace_name`. `Metric` and
+  browser `PageView` have no color dimension (env-level).
+
+**Scenario-UI expectation** is env-derived to match the deploy: `SCENARIO_UI_ENABLED` = `prod` → false,
+else true — so `test_scenario_service.py::test_scenario_ui_flag_controls_webpage` asserts 404 in prod,
+200 elsewhere. Nothing to pass.
+
+---
+
+## Requirements (per environment)
+
+For the **color-aware DB tunnel** the target GitHub Environment needs (Terraform envs already have these):
+- **Variables:** `AKS_CLUSTER_NAME`, `AKS_RESOURCE_GROUP`, `BASE_URL`, `NR_ACCOUNT_ID`
+- **Secrets:** `AZURE_CREDENTIALS` (for `azure/login`), `MSSQL_SA_USER` / `MSSQL_SA_PASSWORD`, `NR_USER_API_KEY`
+
+If `AKS_CLUSTER_NAME` is unset (legacy `events`), the tunnel steps skip and the suite uses the env's own
+`DB_SERVER` / `DB_PASSWORD` secrets against that env's public DB endpoint.
+
+---
+
+## How to run
+
+```bash
+# Manual, against a specific env + color
+gh workflow run test-suite.yml --repo newrelic/relibank --ref <branch> \
+  -f target_environment=sandbox -f test_suite=all -f target_color=blue
+
+# Manual, active color (omit target_color)
+gh workflow run test-suite.yml --repo newrelic/relibank --ref <branch> \
+  -f target_environment=sandbox -f test_suite=all
+
+# Just the scenario-service module (fast)
+gh workflow run test-suite.yml --repo newrelic/relibank --ref <branch> \
+  -f target_environment=sandbox -f test_suite=scenario
+```
+
+**Local** (`skaffold dev`): single-color; leave `TARGET_COLOR` unset and run `cd tests && ./run_tests.sh`.
+The color layers become no-ops (no header, localhost DB, empty NR filters).
+
+---
+
+## Interpreting results
+
+A healthy sandbox run today is **15/17 Python modules green** (plus frontend green). Two modules are a
+**known, pre-existing failure bucket — not color-related and not from any recent change** (they fail the
+same way on `main`'s scheduled runs):
+
+- **`test_newrelic_risk_assessment`** — `NoneType`/`found 0`. New Relic **log/span ingestion lag** (the
+  logs aren't queryable within the test's short wait window) plus non-robust empty-result handling.
+- **`test_nrdot_mssql_collector`** — 5× `Timeout (>300.0s) from pytest-timeout`. NR **metric ingestion**
+  for the mssql collector exceeds the per-test timeout.
+
+Fixing them (longer NR waits + empty-result guards) is tracked separately. Everything else — including
+the color-scoped `test_newrelic_instrumentation` and `test_db_pool_e2e` — should be green.
+
+---
+
+## Gotchas
+
+- **New Relic ingestion lag.** APM/Log/Span/Metric take a harvest + ingest cycle to appear; NR-querying
+  tests generate activity then poll. Transient "found 0" on a warm-then-cold env is usually lag, not a bug.
+- **`kubectl` context.** The DB-tunnel steps run `az aks get-credentials`; when debugging locally, confirm
+  `kubectl config current-context` is the intended sandbox cluster before poking at `relibank-<color>`.
+- **Deploy-time frontend rollout stall (not a test failure).** A full-service roll (e.g. `force_rebuild`,
+  or adding an env to all services) can leave the frontend pod `Pending` on `Insufficient cpu` during
+  surge, which trips Terraform's rollout wait and marks the *deploy* failed even though it converges. See
+  [runbook.md](runbook.md). It surfaces in the `Deploy ReliBank` job, not the test suite.

--- a/k8s/base/databases/mssql-deployment.yaml
+++ b/k8s/base/databases/mssql-deployment.yaml
@@ -71,8 +71,10 @@ spec:
       accessModes:
       - ReadWriteOnce
       resources:
+        # 8Gi headroom; the old 2Gi filled up from unbounded log growth. See init.sql
+        # (RelibankDB is set to SIMPLE recovery so the transaction log self-truncates).
         requests:
-          storage: 2Gi
+          storage: 8Gi
 ---
 apiVersion: v1
 kind: Service

--- a/scenario_service/README.md
+++ b/scenario_service/README.md
@@ -20,6 +20,8 @@ The Scenario Runner Service is a Python-based FastAPI application designed to he
 
 * **Containerized**: The service is packaged in a Docker container for easy deployment in a Kubernetes environment.
 
+* **Environment-gated UI**: The web UI (`/scenario-runner/home`) is hidden in `prod` and shown in every other environment (derived from the deploy environment). See [Web UI visibility](#-web-ui-visibility) below.
+
 ---
 
 ### ⚙️ Getting Started
@@ -42,6 +44,23 @@ To get the Scenario Runner Service up and running, you must deploy it to your Ku
     ```bash
     http://localhost:8000
     ```
+
+---
+
+### 👁️ Web UI visibility
+
+The browser control panel at `/scenario-runner/home` is shown or hidden **based on the deploy
+environment**, while the REST API stays reachable either way (so backend services and automation
+that call `/scenario-runner/api/*` are never affected).
+
+- Driven by the `SCENARIO_UI_ENABLED` env var, which Terraform derives from the environment:
+  **`prod` → `false` (hidden), every other environment → `true` (shown)** —
+  `SCENARIO_UI_ENABLED = tostring(var.demo_environment != "prod")` in the `infrastructure-config`
+  ConfigMap → the scenario-runner pod. There is no per-run flag to set.
+- When `false`, `GET /scenario-runner/home` returns **404** (behaves as if the page doesn't exist).
+- It takes effect on a normal deploy of the target environment; see
+  [docs/deployer/runbook.md → Scenario Runner UI visibility](../docs/deployer/runbook.md#scenario-runner-ui-visibility).
+- Locally, set `SCENARIO_UI_ENABLED` in `skaffold.env`.
 
 ---
 

--- a/scenario_service/scenario_service.py
+++ b/scenario_service/scenario_service.py
@@ -24,6 +24,11 @@ from datetime import datetime, timedelta
 import httpx
 import pyodbc
 
+# Deploy-time flag: whether the browser UI (/scenario-runner/home) is exposed.
+# The scenario API (/scenario-runner/api/*) is always reachable regardless.
+# Read once at startup; toggling requires a redeploy that rolls this pod.
+SCENARIO_UI_ENABLED = os.getenv("SCENARIO_UI_ENABLED", "true").lower() == "true"
+
 # Dictionary to store parsed Chaos Mesh experiments
 CHAOS_EXPERIMENTS: Dict[str, Dict[str, Any]] = {}
 STRESS_EXPERIMENTS: Dict[str, Dict[str, Any]] = {}
@@ -208,7 +213,10 @@ app = FastAPI(title="Relibank Scenario Runner", lifespan=lifespan)
 
 @app.get("/scenario-runner/home", response_class=HTMLResponse)
 async def read_root():
-    """Serves the main HTML page."""
+    """Serves the main HTML page (gated by the SCENARIO_UI_ENABLED deploy flag)."""
+    if not SCENARIO_UI_ENABLED:
+        # Behave as if the page doesn't exist; the API remains reachable.
+        raise HTTPException(status_code=404)
     with open("index.html", "r") as f:
         return f.read()
 

--- a/terraform/aks/app_module/main.tf
+++ b/terraform/aks/app_module/main.tf
@@ -99,6 +99,7 @@ resource "kubernetes_config_map_v1" "infrastructure_config" {
     RISK_ASSESSMENT_SERVICE_URL = "http://risk-assessment-service:5001"
     AZURE_OPENAI_ENDPOINT       = var.azure_openai_endpoint
     ASSISTANT_B_DELAY_SECONDS   = tostring(var.assistant_b_delay_seconds)
+    SCENARIO_UI_ENABLED         = tostring(var.demo_environment != "prod")
   }
 
   depends_on = [kubernetes_namespace_v1.relibank_color]
@@ -214,6 +215,13 @@ resource "kubernetes_deployment_v1" "service" {
 
           port {
             container_port = each.value.container_port
+          }
+
+          # Deploy color on every service — stamped onto NR APM transactions (via the shared
+          # utils/process_headers) so telemetry is filterable by color. One block, all services.
+          env {
+            name  = "DEPLOY_COLOR"
+            value = var.target_color
           }
 
           dynamic "resources" {
@@ -425,7 +433,11 @@ resource "kubernetes_stateful_set_v1" "mssql" {
       spec {
         access_modes = ["ReadWriteOnce"]
         resources {
-          requests = { storage = "2Gi" }
+          # 8Gi headroom for the RelibankDB data file + transaction log. The old 2Gi
+          # filled up (log grew under FULL recovery) and stalled the DB with 1105/1101
+          # allocation errors. RelibankDB is set to SIMPLE recovery in init.sql so the
+          # log self-truncates, but the larger volume gives durable headroom.
+          requests = { storage = "8Gi" }
         }
       }
     }

--- a/terraform/aks/app_module/variables.tf
+++ b/terraform/aks/app_module/variables.tf
@@ -77,6 +77,7 @@ variable "assistant_b_delay_seconds" {
   default     = 0
 }
 
+
 variable "azure_function_url" {
   description = "Full URL (with ?code=) of the notifications Function App. Empty string OK — notifications-service silently no-ops if unset."
   type        = string
@@ -232,6 +233,7 @@ variable "services" {
         DB_SERVER           = "MSSQL_SERVER_NAME"
         DB_DATABASE         = "MSSQL_DATABASE_NAME"
         SUPPORT_SERVICE_URL = "SUPPORT_SERVICE_URL"
+        SCENARIO_UI_ENABLED = "SCENARIO_UI_ENABLED"
       }
       secret_envs = {
         DB_USERNAME = "MSSQL_SA_USER"

--- a/tests/README.md
+++ b/tests/README.md
@@ -182,6 +182,7 @@ All tests support these environment variables for remote testing:
 | `ACCOUNTS_SERVICE` | Accounts service API URL | `http://localhost:5002` |
 | `BILL_PAY_SERVICE` | Bill pay service API URL | `http://localhost:5000` |
 | `SUPPORT_SERVICE` | Support service API URL | `http://localhost:5003` |
+| `TARGET_COLOR` | Deployment color to direct the suite at — sends `X-Test-Env`, port-forwards that color's DB, and scopes New Relic queries. Empty = active color. See [Color-aware testing](#-color-aware-testing-bluegreen). | `""` (active / local single-color) |
 
 ### Automatic Environment Variable Loading (Local Development)
 
@@ -208,6 +209,38 @@ pytest tests/test_newrelic_instrumentation.py
 ```
 
 This approach ensures tests work seamlessly in both local development and CI/CD environments without requiring duplicate environment configuration.
+
+## 🎨 Color-aware testing (blue/green)
+
+Deployed ReliBank environments run **blue/green** — two full stacks (namespaces `relibank-blue` /
+`relibank-green`), one serving live traffic. The suite is **color-aware**: it drives and verifies a
+*specific* color end to end, so a post-deploy run can validate the just-deployed (inactive) color
+**before** traffic is cut to it. All three layers key off one resolved color:
+
+- **HTTP** — [`conftest.py`](conftest.py) attaches an `X-Test-Env: <color>` header to every request
+  (when `TARGET_COLOR` is set), which relibank's canary ingress routes to that color's pods. Empty
+  `TARGET_COLOR` → no header → whatever color is **active**.
+- **Database** — the CI job port-forwards that color's in-cluster MSSQL to `127.0.0.1` and points the
+  DB-direct tests at it (`DB_SERVER=127.0.0.1`). MSSQL stays a private ClusterIP — never publicly exposed.
+- **New Relic** — [`nrql_color.py`](nrql_color.py) `color_filter(event_type)` appends a color `WHERE`
+  clause to NRQL: APM `Transaction` → `deploy.color`, `Span` → `k8s.namespace.name`, `Log` →
+  `namespace_name` (all resolve to the deployed color). `Metric` (mssql-collector) and browser
+  `PageView` have no color dimension and remain env-level.
+
+**Which color** is the *effective color*, resolved once in `test-suite.yml`: the `target_color`
+input if provided (a `Deploy ReliBank` post-deploy run passes the just-deployed color), otherwise the
+**active** color read from `main-ingress`. The scheduled cron passes nothing → validates the active color.
+
+**Scenario-UI expectation** is derived the same way: `test-suite.yml` sets `SCENARIO_UI_ENABLED` from
+`target_environment` (`prod` → off, else on), matching how Terraform derives the deployed value, so
+`test_scenario_service.py::test_scenario_ui_flag_controls_webpage` asserts the right state per env.
+
+**Local runs** (`skaffold dev` / `run_tests.sh`) are single-color — leave `TARGET_COLOR` unset and these
+layers are no-ops (no header; `DB_SERVER` defaults to localhost; NR filters render empty).
+
+See the **[Testing Runbook](../docs/deployer/testing-runbook.md)** for how to run the suite against a
+specific environment/color, DB access requirements, and how to interpret results (including the known
+pre-existing failures).
 
 ## Troubleshooting
 

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1,0 +1,45 @@
+"""Shared pytest fixtures for the relibank test suite.
+
+Color-directed testing (modeled on demogorgon): when the ``TARGET_COLOR`` env var is set,
+every HTTP request the suite makes carries an ``X-Test-Env: <color>`` header so it routes
+through relibank's canary ingress to that specific deployment color. The active color is
+reached whether or not the header is present, so an empty ``TARGET_COLOR`` (e.g. the
+scheduled cron run) simply exercises whatever color is live.
+
+This is applied centrally by patching ``requests.Session.request`` in an autouse fixture,
+so individual test modules need no changes — every ``requests.get``/``post`` (which use a
+transient Session under the hood) picks it up. Adding the header to non-relibank calls
+(e.g. the New Relic NerdGraph API) is harmless; unknown headers are ignored.
+"""
+import os
+
+import pytest
+import requests
+
+
+@pytest.fixture(scope="session")
+def target_color():
+    """Deployment color the run is directed at; '' means active color / no routing."""
+    return os.getenv("TARGET_COLOR", "").strip()
+
+
+@pytest.fixture(autouse=True, scope="session")
+def _route_requests_to_color(target_color):
+    """Attach ``X-Test-Env: <color>`` to every request when a color is directed."""
+    if not target_color:
+        yield
+        return
+
+    original_request = requests.sessions.Session.request
+
+    def request_with_color(self, method, url, **kwargs):
+        headers = dict(kwargs.get("headers") or {})
+        headers.setdefault("X-Test-Env", target_color)  # explicit per-call header wins
+        kwargs["headers"] = headers
+        return original_request(self, method, url, **kwargs)
+
+    requests.sessions.Session.request = request_with_color
+    try:
+        yield
+    finally:
+        requests.sessions.Session.request = original_request

--- a/tests/nrql_color.py
+++ b/tests/nrql_color.py
@@ -1,0 +1,54 @@
+"""Color-directed NRQL helper.
+
+relibank encodes the deploy color in the Kubernetes namespace (``relibank-<color>``), which
+``nri-metadata-injection`` stamps onto APM/Log/Span telemetry — under an event-type-specific
+attribute name. This is the relibank analog of demogorgon filtering NRQL on the color-bearing
+pod/host name.
+
+Usage — drop it into a ``WHERE`` so the query reads naturally:
+
+    f"... AND db.pool_id IS NOT NULL {color_filter('Transaction')} SINCE 10 minutes ago"
+
+It renders to ``AND `namespaceName` = 'relibank-blue'`` when a color is directed, or an empty
+string when ``TARGET_COLOR`` is unset (e.g. the scheduled cron resolving to the active color
+with no directive) — so the query stays env-level.
+
+Not all telemetry carries the namespace: the mssql-collector ``Metric`` stream and browser
+events (``PageView``/``BrowserInteraction``, shared browser app) have no namespace dimension,
+so they remain color-blind by necessity — do not use this helper for them.
+"""
+import os
+
+# (attribute, value-template) carrying the color, per NR event type. The value template is
+# formatted with the effective color. Two flavors:
+#   - APM Transaction/TransactionError: the ``deploy.color`` custom attribute (value = the bare
+#     color, e.g. "blue"), stamped by utils/process_headers.py. nri-metadata-injection does NOT
+#     put a namespace on APM events, so this custom attribute is how APM is color-scoped.
+#   - Span/Log: the k8s namespace nri-metadata-injection stamps (value = "relibank-<color>"):
+#     Span -> k8s.namespace.name, Log -> namespace_name.
+# Metric (mssql collector) and browser PageView have no color dimension and are intentionally
+# absent -> color_filter() is a no-op for them (they stay env-level).
+_COLOR_DIM = {
+    "Transaction": ("deploy.color", "{color}"),
+    "TransactionError": ("deploy.color", "{color}"),
+    "Span": ("k8s.namespace.name", "relibank-{color}"),
+    "Log": ("namespace_name", "relibank-{color}"),
+}
+
+
+def effective_color():
+    """The color the run is directed at ('' = undirected / active, no filtering)."""
+    return os.getenv("TARGET_COLOR", "").strip()
+
+
+def color_filter(event_type):
+    """NRQL ``AND`` fragment scoping to the effective color's namespace, or '' if undirected.
+
+    Designed to sit inline after an existing predicate, e.g. ``... IS NOT NULL {color_filter('Span')}``.
+    """
+    color = effective_color()
+    dim = _COLOR_DIM.get(event_type)
+    if not color or not dim:
+        return ""
+    attr, value_tmpl = dim
+    return f"AND `{attr}` = '{value_tmpl.format(color=color)}'"

--- a/tests/test_db_pool_e2e.py
+++ b/tests/test_db_pool_e2e.py
@@ -10,6 +10,8 @@ import os
 import hashlib
 from pathlib import Path
 
+from nrql_color import color_filter
+
 # Helper function to load environment variables from skaffold.env if present
 def load_env_from_skaffold():
     """Load environment variables from skaffold.env if file exists (local development)"""
@@ -178,11 +180,12 @@ def test_db_pool_e2e_with_new_relic_validation(reset_scenario):
     print("\n📝 Step 5: Validating custom attributes in New Relic...")
 
     # Query for db.pool_id (custom attributes don't get custom. prefix when using add_custom_attribute)
-    nrql = """
+    nrql = f"""
     SELECT count(*), uniques(db.pool_id)
     FROM Transaction
     WHERE appName LIKE '%Accounts Service%'
       AND db.pool_id IS NOT NULL
+      {color_filter('Transaction')}
     SINCE 10 minutes ago
     """
 
@@ -204,13 +207,14 @@ def test_db_pool_e2e_with_new_relic_validation(reset_scenario):
     # Step 6: Validate performance difference
     print("\n📝 Step 6: Validating performance difference between pools...")
 
-    perf_nrql = """
+    perf_nrql = f"""
     SELECT
       average(duration) as 'avg_duration',
       count(*) as 'count'
     FROM Transaction
     WHERE appName LIKE '%Accounts Service%'
       AND db.pool_id IS NOT NULL
+      {color_filter('Transaction')}
     FACET db.pool_id
     SINCE 10 minutes ago
     """

--- a/tests/test_newrelic_instrumentation.py
+++ b/tests/test_newrelic_instrumentation.py
@@ -19,6 +19,8 @@ import requests
 from datetime import datetime, timedelta
 from pathlib import Path
 
+from nrql_color import color_filter
+
 # Helper function to load environment variables from skaffold.env if present
 def load_env_from_skaffold():
     """Load environment variables from skaffold.env if file exists (local development)"""
@@ -230,7 +232,7 @@ def test_newrelic_transactions_exist():
     print("\n=== Test: New Relic Transactions Exist ===")
 
     # Query for recent transactions (last 5 minutes)
-    nrql = "SELECT count(*) FROM Transaction WHERE appName LIKE '%Relibank%' OR appName LIKE '%Accounts Service%' OR appName LIKE '%Transaction Service%' OR appName LIKE '%Bill Pay%' SINCE 5 minutes ago"
+    nrql = f"SELECT count(*) FROM Transaction WHERE (appName LIKE '%Relibank%' OR appName LIKE '%Accounts Service%' OR appName LIKE '%Transaction Service%' OR appName LIKE '%Bill Pay%') {color_filter('Transaction')} SINCE 5 minutes ago"
 
     print(f"Querying NRQL: {nrql}")
     results = query_nrql(nrql)
@@ -258,7 +260,7 @@ def test_newrelic_user_id_tracking():
         pytest.skip("Could not generate test traffic - services may not be accessible")
 
     # Query for transactions with our test user ID
-    nrql = f"SELECT count(*) FROM Transaction WHERE enduser.id = '{test_user_id}' SINCE 5 minutes ago"
+    nrql = f"SELECT count(*) FROM Transaction WHERE enduser.id = '{test_user_id}' {color_filter('Transaction')} SINCE 5 minutes ago"
 
     print(f"\nQuerying NRQL: {nrql}")
     results = query_nrql(nrql)
@@ -318,7 +320,7 @@ def test_newrelic_services_instrumented():
 
     for service in expected_services:
         # Check for transactions from this service in the last hour
-        nrql = f"SELECT count(*) FROM Transaction WHERE appName LIKE '%{service}%' SINCE 1 hour ago"
+        nrql = f"SELECT count(*) FROM Transaction WHERE appName LIKE '%{service}%' {color_filter('Transaction')} SINCE 1 hour ago"
 
         print(f"\nChecking {service}...")
         print(f"  Query: {nrql}")
@@ -353,7 +355,7 @@ def test_newrelic_recent_activity():
     """Verify there has been recent activity in the last 10 minutes."""
     print("\n=== Test: Recent Activity ===")
 
-    nrql = "SELECT count(*) FROM Transaction WHERE appName LIKE '%Relibank%' OR appName LIKE '%Accounts Service%' OR appName LIKE '%Transaction Service%' OR appName LIKE '%Bill Pay%' SINCE 10 minutes ago"
+    nrql = f"SELECT count(*) FROM Transaction WHERE (appName LIKE '%Relibank%' OR appName LIKE '%Accounts Service%' OR appName LIKE '%Transaction Service%' OR appName LIKE '%Bill Pay%') {color_filter('Transaction')} SINCE 10 minutes ago"
 
     print(f"Querying NRQL: {nrql}")
     results = query_nrql(nrql)
@@ -378,7 +380,7 @@ def test_newrelic_error_notice_tracking():
         "WHERE (appName LIKE '%Relibank%' OR appName LIKE '%Bill Pay%') "
         "AND custom.service IS NOT NULL "
         "AND custom.endpoint IS NOT NULL "
-        "AND custom.action IS NOT NULL "
+        f"AND custom.action IS NOT NULL {color_filter('TransactionError')} "
         "SINCE 10 minutes ago"
     )
 
@@ -410,7 +412,7 @@ def test_newrelic_transaction_custom_attributes():
     nrql = (
         "SELECT count(*) FROM Transaction "
         "WHERE appName LIKE '%Bill Pay%' "
-        "AND billId IS NOT NULL "
+        f"AND billId IS NOT NULL {color_filter('Transaction')} "
         "SINCE 10 minutes ago"
     )
 

--- a/tests/test_newrelic_risk_assessment.py
+++ b/tests/test_newrelic_risk_assessment.py
@@ -22,6 +22,8 @@ import pyodbc
 from pathlib import Path
 from typing import Dict, List
 
+from nrql_color import color_filter
+
 # Helper function to load environment variables from skaffold.env if present
 def load_env_from_skaffold():
     """Load environment variables from skaffold.env if file exists (local development)"""
@@ -243,10 +245,11 @@ def test_risk_assessment_logs_in_newrelic(reset_risk_scenarios):
     nrql = f"""
     SELECT count(*) as log_count, latest(message) as sample_message
     FROM Log
-    WHERE (message LIKE '%risk%' OR message LIKE '%assess%')
+    WHERE ((message LIKE '%risk%' OR message LIKE '%assess%')
       AND entity.name IN ('{APP_NAME_PREFIX} - Bill Pay Service',
                           '{APP_NAME_PREFIX} - Support Service',
-                          '{APP_NAME_PREFIX} - Risk Assessment Service')
+                          '{APP_NAME_PREFIX} - Risk Assessment Service'))
+      {color_filter('Log')}
     SINCE 2 minutes ago
     """
 
@@ -325,9 +328,10 @@ def test_declined_payment_shows_in_newrelic(reset_risk_scenarios):
     nrql = f"""
     SELECT count(*) as declined_log_count
     FROM Log
-    WHERE message LIKE '%declined%' OR message LIKE '%DECLINED%'
+    WHERE (message LIKE '%declined%' OR message LIKE '%DECLINED%'
       AND entity.name IN ('{APP_NAME_PREFIX} - Bill Pay Service',
-                          '{APP_NAME_PREFIX} - Support Service')
+                          '{APP_NAME_PREFIX} - Support Service'))
+      {color_filter('Log')}
     SINCE 3 minutes ago
     """
 
@@ -446,8 +450,9 @@ def test_agent_model_tracking_in_newrelic(reset_risk_scenarios):
     nrql = f"""
     SELECT count(*) as log_count
     FROM Log
-    WHERE (message LIKE '%gpt-4o%' OR message LIKE '%agent_model%')
-      AND entity.name = '{APP_NAME_PREFIX} - Support Service'
+    WHERE ((message LIKE '%gpt-4o%' OR message LIKE '%agent_model%')
+      AND entity.name = '{APP_NAME_PREFIX} - Support Service')
+      {color_filter('Log')}
     SINCE 3 minutes ago
     """
 
@@ -502,9 +507,10 @@ def test_risk_assessment_ebpf_traces(reset_risk_scenarios):
     nrql = f"""
     SELECT count(*) as span_count, average(duration) as avg_duration_ms
     FROM Span
-    WHERE entity.name = '{APP_NAME_PREFIX} - Risk Assessment Service'
+    WHERE (entity.name = '{APP_NAME_PREFIX} - Risk Assessment Service'
       OR name LIKE '%assess-risk%'
-      OR name LIKE '%risk-assessment%'
+      OR name LIKE '%risk-assessment%')
+      {color_filter('Span')}
     SINCE 3 minutes ago
     """
 
@@ -527,6 +533,7 @@ def test_risk_assessment_ebpf_traces(reset_risk_scenarios):
           AND (risk_level IS NOT NULL
                OR decision IS NOT NULL
                OR agent_model IS NOT NULL)
+          {color_filter('Span')}
         SINCE 3 minutes ago
         """
 

--- a/tests/test_scenario_service.py
+++ b/tests/test_scenario_service.py
@@ -20,16 +20,63 @@ def reset_scenarios_after_test():
 
 
 def test_scenario_service_health():
-    """Test that scenario service is accessible"""
+    """Test that scenario service is accessible.
+
+    Probes the always-on /scenario-runner health route rather than the UI page
+    (/scenario-runner/home), which is intentionally 404 when the SCENARIO_UI_ENABLED
+    deploy flag is off. Health must not be coupled to UI visibility.
+    """
     print("\n=== Testing Scenario Service Health ===")
 
     try:
-        response = requests.get(f"{SCENARIO_SERVICE_URL}/scenario-runner/home", timeout=5)
+        response = requests.get(f"{SCENARIO_SERVICE_URL}/scenario-runner", timeout=5)
         print(f"Status: {response.status_code}")
         assert response.status_code == 200, f"Scenario service not accessible: {response.status_code}"
         print("✓ Scenario service is accessible")
     except requests.exceptions.ConnectionError:
         pytest.fail("Cannot connect to scenario service")
+
+
+def test_scenario_ui_flag_controls_webpage():
+    """Verify the browser UI page visibility matches the SCENARIO_UI_ENABLED deploy flag.
+
+    Color-directed (demogorgon convention): when TARGET_COLOR is set, every request carries
+    an `X-Test-Env: <color>` header so it routes to that specific color (post-deployment runs
+    always set it, so the just-deployed color is validated before cutover); when unset, the
+    active color is hit. The expected state comes from SCENARIO_UI_ENABLED, threaded from the
+    same deploy that set it (default "true" — the flag's own default).
+
+    Asserts:
+      - /scenario-runner/home  -> 200 + page marker when enabled, else 404
+      - /scenario-runner/api/scenarios -> 200 regardless (the flag's invariant: API stays up)
+    """
+    print("\n=== Testing Scenario UI Flag Controls Webpage ===")
+
+    target_color = os.getenv("TARGET_COLOR", "").strip()
+    headers = {"X-Test-Env": target_color} if target_color else {}
+    # Treat unset OR empty (e.g. scheduled runs with no workflow input) as the flag's default (true).
+    expected_enabled = (os.getenv("SCENARIO_UI_ENABLED") or "true").strip().lower() == "true"
+    print(f"target_color={target_color or '<active>'} expected_enabled={expected_enabled}")
+
+    # The webpage page (gated by the flag)
+    home = requests.get(f"{SCENARIO_SERVICE_URL}/scenario-runner/home", headers=headers, timeout=10)
+    print(f"/scenario-runner/home -> {home.status_code}")
+    if expected_enabled:
+        assert home.status_code == 200, (
+            f"UI enabled but /scenario-runner/home returned {home.status_code} (expected 200)"
+        )
+        assert "Relibank Scenario Runner" in home.text, "UI page served but missing expected content marker"
+    else:
+        assert home.status_code == 404, (
+            f"UI disabled but /scenario-runner/home returned {home.status_code} (expected 404)"
+        )
+
+    # The API must remain reachable regardless of the UI flag (core invariant)
+    api = requests.get(f"{SCENARIO_SERVICE_URL}/scenario-runner/api/scenarios", headers=headers, timeout=10)
+    print(f"/scenario-runner/api/scenarios -> {api.status_code}")
+    assert api.status_code == 200, f"Scenario API not reachable ({api.status_code}) — should be up regardless of UI flag"
+
+    print(f"✓ Webpage visibility matches SCENARIO_UI_ENABLED={expected_enabled}; API reachable")
 
 
 def test_get_all_scenarios():

--- a/transaction_service/mssql/init.sql
+++ b/transaction_service/mssql/init.sql
@@ -27,6 +27,13 @@ BEGIN
 END;
 GO
 
+-- Use SIMPLE recovery so the transaction log self-truncates on checkpoint.
+-- This is a demo/sandbox DB re-seeded on each fresh deploy; there is no log-backup
+-- job, so FULL recovery caused the log (.ldf) to grow without bound and fill the
+-- data volume, stalling the server with 1105/1101 allocation errors. Idempotent.
+ALTER DATABASE RelibankDB SET RECOVERY SIMPLE;
+GO
+
 -- ==============================================================================
 -- NEW RELIC MONITORING SETUP - SERVER LEVEL
 -- ==============================================================================

--- a/utils/process_headers.py
+++ b/utils/process_headers.py
@@ -1,9 +1,15 @@
+import os
 import time
 import logging
 from typing import Dict, Any
 
 from fastapi import HTTPException
 import newrelic.agent
+
+# Deploy color (blue/green), injected per-pod via the DEPLOY_COLOR env var. Stamped as a
+# custom attribute on every transaction so New Relic APM events can be filtered by color
+# (nri-metadata-injection surfaces the namespace onto Logs/Spans but NOT APM Transactions).
+DEPLOY_COLOR = os.getenv("DEPLOY_COLOR", "")
 
 # This function is now synchronous (def) and uses time.sleep(),
 # which WILL BLOCK the entire application process.
@@ -20,6 +26,13 @@ def process_headers(headers: Dict[str, str | Any]):
     Args:
         headers: A dictionary (from Request.headers) where keys are all lowercase.
     """
+
+    # Stamp the deploy color on every transaction so APM events are filterable by color.
+    if DEPLOY_COLOR:
+        try:
+            newrelic.agent.add_custom_attribute("deploy.color", DEPLOY_COLOR)
+        except Exception as e:
+            logging.warning(f"[Color Tagging] Failed to set deploy.color: {e}")
 
     # 0. Handle New Relic APM user tracking
     browser_user_id = headers.get("x-browser-user-id")

